extra-tests/classes/RollupQueryBuilderTests.cls

```diff
@@ -74,6 +74,25 @@ private class RollupQueryBuilderTests {
     System.assertEquals(Event.SObjectType, records.getSObjectType());
   }
 
+  @IsTest
+  static void handlesNestedPolymorphicWhereClausesWithOr() {
+    // we expect that What.Type will end up being stripped out of this
```

Comment:
What's the reason that `What.Type` is expected to be stripped from the where clause? Based on your `RollupQueryBuilder` class, it seems like `What.Type` would stay (because it's valid SOQL), but `What.Owner.Id` would be stripped (because it's not valid SOQL), but I'm probably misreading something here

Reply:
they both end up being parsed out in favor of a `TYPE OF` selection.